unicorn/include/cocaine/zookeeper/handler.hpp

```diff
@@ -0,0 +1,230 @@
+
+    static
+    void
+        string_cb(int rc, const char *value, const void *data);
+
+    static
+    void
+        strings_stat_cb(int rc, const struct String_vector *strings, const struct Stat *stat, const void *data);
+
+private:
+    typedef std::shared_ptr<managed_handler_base_t> handler_ptr;
+
+    //Ugly hack. That is fixed in C++14 - we can search in set by convertible value. Now we use map.
+    typedef std::unordered_map<const managed_handler_base_t*, handler_ptr> storage_t;
+
+    friend
+    class handler_scope_t;
+
+    static handler_dispatcher_t& instance();
+
+    template<class T, class ...Args>
+    void
+    call(const void* callback, Args&& ...args) {
+        handler_ptr cb;
+        {
+            std::unique_lock<std::mutex> lock(storage_lock);
+            auto it = callbacks.find(reinterpret_cast<const managed_handler_base_t*>(callback));
+            if (it != callbacks.end()) {
+                cb = it->second;
+                assert(cb);
+            }
+        }
+        if(cb) {
+            std::dynamic_pointer_cast<T>(cb)->operator()(std::forward<Args>(args)...);
+        }
+    }
+
+    void add(managed_handler_base_t* callback);
+
+    void release(managed_handler_base_t* callback);
+
+    handler_dispatcher_t() {}
+    handler_dispatcher_t(const handler_dispatcher_t& other) = delete;
+    handler_dispatcher_t& operator=(const handler_dispatcher_t& other) = delete;
+    std::mutex storage_lock;
+    storage_t callbacks;
+};
+
+/**
+* Scope of the handlers. After this goes out of scope all handlers ae destroyed.
+* However call of this handlers via handler manager is defined and do nothing.
+* Thread unsafe. For thread safety consider external locking or creating separate class.
+*/
+class handler_scope_t {
+public:
+    handler_scope_t() :
+        registered_callbacks()
+    {}
+
+    ~handler_scope_t();
+
+    template<class T, class ...Args>
+    T&
+    get_handler(Args&& ...args) {
+        auto handler = new T(handler_tag(), std::forward<Args>(args)...);
+        handler_dispatcher_t::instance().add(handler);
+        registered_callbacks.push_back(handler);
+        return *handler;
+    }
+
+private:
+    std::vector<managed_handler_base_t*> registered_callbacks;
+};
+
+class managed_watch_handler_base_t :
+    virtual public managed_handler_base_t
+{
+public:
+    virtual void
+        operator() (int type, int state, path_t path) = 0;
+
+    managed_watch_handler_base_t(const handler_tag& tag) :
+        managed_handler_base_t(tag)
+    {}
+};
+
+/**
+* At this time managed void handler is unneeded so it's being deleted after invocation.
+* Maybe better to stick to only managed handlers for consistency.
+*/
+class void_handler_base_t {
+public:
+    virtual void
+    operator() (int rc) = 0;
+
+    virtual
+    ~void_handler_base_t() {}
+};
+
+class managed_stat_handler_base_t :
+    virtual public managed_handler_base_t
```

Comment:
We need to store all calbacks as base type in dispatcher. However some class would want to implement several callbacks at a time(this seems reasonable if they share same context) in this case we have diamond inheritance.